blockchain/state_manager/src/lib.rs

```diff
@@ -313,64 +406,47 @@ where
-
-        let cid_pair = if tipset.epoch() == 0 {
-            // NB: This is here because the process that executes blocks requires that the
-            // block miner reference a valid miner in the state tree. Unless we create some
-            // magical genesis miner, this won't work properly, so we short circuit here
-            // This avoids the question of 'who gets paid the genesis block reward'
-            let message_receipts = tipset
-                .blocks()
-                .first()
-                .ok_or_else(|| Error::Other("Could not get message receipts".to_string()))?;
-
-            (*tipset.parent_state(), *message_receipts.message_receipts())
-        } else {
-            // generic constants are not implemented yet this is a lowcost method for now
-            let no_func = None::<fn(&Cid, &ChainMessage, &ApplyRet) -> Result<(), anyhow::Error>>;
-            let ts_state = self.compute_tipset_state(tipset, no_func).await?;
-            debug!("completed tipset state calculation {:?}", tipset.cids());
-            ts_state
-        };
-
-        // Fill entry with calculated cid pair
-        if let Err(e) = cache_entry.set(cid_pair) {
-            error!("Fail to set tipset_state cache: {}, {}", e.0, e.1);
-        }
-        forest_metrics::metrics::LRU_CACHE_MISS
-            .with_label_values(&[forest_metrics::metrics::values::STATE_MANAGER_TIPSET])
-            .inc();
-        Ok(cid_pair)
+        //let mut cache_miss = false;
+        let key = tipset.key();
+        let result = self
+            .cache
+            .get_or_else(key, |key| async move {
+                //cache_miss = true;
+                let cid_pair = if tipset.epoch() == 0 {
+                    // NB: This is here because the process that executes blocks requires that the
+                    // block miner reference a valid miner in the state tree. Unless we create some
+                    // magical genesis miner, this won't work properly, so we short circuit here
+                    // This avoids the question of 'who gets paid the genesis block reward'
+                    let message_receipts = tipset.blocks().first().ok_or_else(|| {
+                        Error::Other("Could not get message receipts".to_string())
+                    })?;
+
+                    (*tipset.parent_state(), *message_receipts.message_receipts())
+                } else {
+                    // generic constants are not implemented yet this is a lowcost method for now
+                    let no_func =
+                        None::<fn(&Cid, &ChainMessage, &ApplyRet) -> Result<(), anyhow::Error>>;
+                    let ts_state = self.compute_tipset_state(tipset, no_func).await?;
+                    debug!("Completed tipset state calculation {:?}", tipset.cids());
+                    ts_state
+                };
+
+                // Write back to cache, release lock and return value
+                self.cache.insert(key.clone(), cid_pair);
```

Comment:
Shouldn't this be done in `get_or_else`?

Reply:
Indeed, I was hesitating to move it here, will do.